pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocator.java

```diff
@@ -0,0 +1,278 @@
+
+  private final ScheduledExecutorService _executorService;
+  private final PinotHelixResourceManager _pinotHelixResourceManager;
+  private final HelixManager _helixManager;
+  private final HelixAdmin _helixAdmin;
+  private final long _runFrequencySeconds;
+
+  public RealtimeSegmentRelocator(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf config) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _helixManager = pinotHelixResourceManager.getHelixZkManager();
+    _helixAdmin = pinotHelixResourceManager.getHelixAdmin();
+    _runFrequencySeconds = getRunFrequencySeconds(config.getRealtimeSegmentRelocatorFrequency());
+
+    _executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+      @Override
+      public Thread newThread(Runnable runnable) {
+        Thread thread = new Thread(runnable);
+        thread.setName("RealtimeSegmentRelocatorExecutorService");
+        return thread;
+      }
+    });
+  }
+
+  public void start() {
+    LOGGER.info("Starting realtime segment relocator");
+
+    _executorService.scheduleWithFixedDelay(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          runRelocation();
+        } catch (Exception e) {
+          LOGGER.warn("Caught exception while running realtime segment relocator", e);
+        }
+      }
+    }, 120, _runFrequencySeconds, TimeUnit.SECONDS);
+  }
+
+  public void stop() {
+    _executorService.shutdown();
+  }
+
+  /**
+   * Check all tables. Perform relocation of segments if table is realtime and relocation is required
+   */
+  public void runRelocation() {
+    if (!_pinotHelixResourceManager.isLeader()) {
+      LOGGER.info("Skipping realtime segment relocation, not leader!");
+      return;
+    }
+
+    LOGGER.info("Starting relocation of realtime segments");
+    List<String> allRealtimeTableNames = _pinotHelixResourceManager.getAllRealtimeTables();
+
+    for (final String tableNameWithType : allRealtimeTableNames) {
+      try {
+        LOGGER.info("Starting relocation of segments for table: {}", tableNameWithType);
+
+        TableConfig tableConfig = _pinotHelixResourceManager.getRealtimeTableConfig(tableNameWithType);
+        final RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        if (!realtimeTagConfig.isRelocateCompletedSegments()) {
+          LOGGER.info("Skipping relocation of segments for {}", tableNameWithType);
+          return;
+        }
+
+        Function<IdealState, IdealState> updater = new Function<IdealState, IdealState>() {
+          @Nullable
+          @Override
+          public IdealState apply(@Nullable IdealState idealState) {
+            if (!idealState.isEnabled()) {
+              LOGGER.info("Skipping relocation of segments for {} since ideal state is disabled", tableNameWithType);
+              return null;
+            }
+            relocateSegments(realtimeTagConfig, idealState);
```

Comment:
its probably better to clone the idealstate and then compare if there was a change before updating the idealstate. Currently, we might just keep updating the idealstate everytime this runs.

Reply:
All that happens inside HelixHelper.updateIdealState. It compares updated ideal state and original ideal state, and updates only if there's a change